### PR TITLE
chore(docs): ACS API guidelines - Part 1: Naming convention

### PR DESCRIPTION
**Pull request by md2119**
## Description

This PR is the first in the series of PRs to document general ACS API best practices. This PR exclusively focuses on the naming conventions recommendations.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed
n/a

### CODE REVIEW

File: .github/api_best_practices.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
+# ACS API Guidelines
```
Comment (msugakov) [inline .github/api_best_practices.md:1]: Please mention kinds of APIs this document focuses because they are different. REST, gRPC, GraphQL, Kubernetes (CRDs).

File: .github/api_best_practices.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+e.g. `ReportConfigService`, `RbacService`. 
+All RPC methods grouped into a single service must generally pertain to the primary resource of the service.
+
+All service defined in versioned package `/proto/api/v2` **must not** import from 
+`/proto/storage` and `/proto/internalapi` packages.
+All new RPC methods defined in versioned package `/proto/api/v1` **must not** import from
+`/proto/storage` and `/proto/internalapi` packages.
+
+### Method Name
+Methods (and URLs) should be named such that they provide insights into the functionality. Typically, the method
```
Comment (vjwilson) [inline .github/api_best_practices.md:30]: This is the only mention of URLs in this doc.

I would like us to specify that URLs should use hyphens to separate words in the URL.

It makes it easier to parse the URL quickly, and also reduces the chance of typos for users who are using the API at the command line.

You can see this in the GitHub API here, with the `dependency-graph` part of the URL,
https://docs.github.com/en/rest/dependency-graph/dependency-review?apiVersion=2022-11-28#get-a-diff-of-the-dependencies-between-commits

It's also part of Google's URL style guide for human-readable URLs,
https://developers.google.com/search/docs/crawling-indexing/url-structure

**Review by vjwilson — COMMENTED** [review]
One comment asking for a statement on URLs

**Review by msugakov — CHANGES_REQUESTED** [review]
I might provide more comments but that :arrow_up: one for me is a blocker. We need to be clear info about which kinds of APIs the reader will find here and for which ones - not.

File: .github/api_best_practices.md (hunk 2 above)
Comment (md2119) [reply .github/api_best_practices.md:30]: @vjwilson good catch. The section for URL naming will be separate. I missed removing this one reference to URLs.

File: .github/api_best_practices.md (hunk 1 above)
Comment (md2119) [reply .github/api_best_practices.md:1]: This document intends to provide guidelines for all APIs, which will come in separate PRs. I don't see the point of enlisting each category here, for that there is a table of contents.

Comment (msugakov) [reply .github/api_best_practices.md:1]: My point is in similar direction as Van's comment about URLs. The following section `Naming Guidelines` talks about gRPC and proto naming. How does the reader know that? Well, some amount of experience will allow inferring that context from the content itself, but possessing that context is already high enough bar.

I think, this bar is set higher than necessary because we'd want less experienced engineers to also benefit from reading these guidelines. Anyone can read anything, the question here is that we make it harder to comprehend.

The second, related, part of my point is that the presented naming guidelines make little sense for Kubernetes APIs which are mostly resource objects (nouns) instead of actions (verbs). If someone wants to put guidelines on Kubernetes APIs naming, how does that person know if this document is fit for that?

The last part of my point is that by making it explicit which API kinds will be in the document, you'd set expectations for reviewers about what they'll find in the future iterations. You'd also be able to source info from reviewers experience about important differences if certain guidelines don't apply for all kinds.

My honest fear is that the guidelines are going to default to gRPC and the rest will be an afterthought, at best. It's fine to address gRPC APIs only but let's be explicit about that. The opposite of doing gRPC with few touches about REST, GraphQL and/or Kubernetes would leave the first impression of something thorough but makes bitter taste when applied in practice because the other kinds are substantially different in some aspects to be covered in a manner of deviations from/extensions of gRPC.

File: .github/api_best_practices.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+Service name **must** be unique and should use a noun that generally refers to a resource or product component 
+e.g. `DeploymentService`, `ReportService`, `ComplianceService`. Intuitive and well-known short forms or 
+abbreviations may be used in some cases (and could even be preferable) for succinctness 
+e.g. `ReportConfigService`, `RbacService`. 
+All RPC methods grouped into a single service must generally pertain to the primary resource of the service.
+
+All service defined in versioned package `/proto/api/v2` **must not** import from 
+`/proto/storage` and `/proto/internalapi` packages.
+All new RPC methods defined in versioned package `/proto/api/v1` **must not** import from
+`/proto/storage` and `/proto/internalapi` packages.
```
Comment (msugakov) [inline .github/api_best_practices.md:27]: This paragraph feels to not belong to Service Naming. Rather, this talks about proto files layout in repo and import rules.
I suggest moving it to a separate section, even if it's just declared as a stub. We'll be able to rename and move when the text is in a separate section, but we'll likely forget to move it after a couple iterations if this text stays here.

File: .github/api_best_practices.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+<tr>
+<td>
+
+`GetDeploymentWithImageScanResponse`
+
+</td>
+<td>
+
+```
+GetDeploymentWithImageScanResponse
```
Comment (msugakov) [inline .github/api_best_practices.md:166]: Is there no mistake to have `GetDeploymentWithImageScanResponse` second time here?

File: .github/api_best_practices.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+GetDeploymentResponse {
+  Deployment deployment;
+  Image image;
+}
```
Comment (msugakov) [inline .github/api_best_practices.md:170]: Not exactly in scope here, rather an inspired question. Does it happen that we need a deployment coupled with different other objects in _different_ response types? What's the recommendation then, I suppose we shouldn't create a mega response type with multiple optional attributes, right? A completely made-up example:
```
GetDeplomentResponse {
  Deployment deployment;
  Image image; // Will be set by GetDeploymentImageScan
  NetworkPolicy network_policy; // Will be set by InferDeploymentNetworkPolicy
  ProcessBaseline process_baseline; // Will be set by GetDeploymentProcessBaseline
}
```

I suppose we'll create multiple Response types, one per operation. Can we reuse `GetDeploymentResponse` name if it looks tempting?

**Review by msugakov — COMMENTED** [review]
The guidelines make sense and read quite well.

File: .github/api_best_practices.md (hunk 1 above)
Comment (md2119) [reply .github/api_best_practices.md:1]: > The following section Naming Guidelines talks about gRPC and proto naming. How does the reader know that?

Fair point. The individual section title can be expanded like `Method Name` -> `RPC Method Name` and likewise. I will do that. 

I want to mention the PRs will be much smaller chunks, therefore, we should refrain from asking for things that are not the PR, and rather focus on things that are part of the PR.

File: .github/api_best_practices.md (hunk 5 above)
Comment (md2119) [reply .github/api_best_practices.md:170]: Your suggestions are completely valid and I will add them to examples.

I have come across customers who demand mega responses. Now, the mega response that we both are talking about could be different, are we talking about every storage field in the response? Not necessarily, but that is a separate topic about decoupling storage and APIs.

File: .github/api_best_practices.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+e.g. `ReportConfigService`, `RbacService`. 
+All RPC methods grouped into a single service must generally pertain to the primary resource of the service.
+
+All service defined in versioned package [/proto/api/v2](https://github.com/stackrox/stackrox/blob/master/proto/v2)
+**must not** import from [/proto/storage](https://github.com/stackrox/stackrox/blob/master/proto/storage) and
+[/proto/internalapi](https://github.com/stackrox/stackrox/blob/master/proto/internalapi) packages.
+All new RPC methods defined in versioned package [/proto/api/v1](https://github.com/stackrox/stackrox/blob/master/proto/v1) **must not** import from
+`/proto/storage` and `/proto/internalapi` packages.
+
+### RPC Method Name
```
Comment (msugakov) [inline .github/api_best_practices.md:30]: (*)

Caring about less experienced readers, I suggest to stick with `gRPC` instead of "generic" `RPC` which can create confusion.

```suggestion
### gRPC Method Name
```

Please also search-replace `RPC` with `gRPC` in the entire doc.

File: .github/api_best_practices.md (hunk 3 above)
Comment (msugakov) [reply .github/api_best_practices.md]: These are guidelines but these aren't **naming** guidelines.

Anyhow, I don't feel pedantic enough to fight for that so it can stay here.

**Review by msugakov — COMMENTED** [review]
Only two new comments I marked with (*) are currently blockers for me. Since these are cosmetic, should be easily addressable, I think.

File: .github/api_best_practices.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+## Table of Contents
+
+- [Naming Guidelines](#naming-guidelines)
+  - [Service Name](#service-name)
+  - [RPC Method Name](#rpc-method-name)
+  - [RPC Message Name](#rpc-message-name)
+
+## Naming Guidelines
+
+### Service Name
```
[the remark this answers is not in the document]
Comment (md2119) [reply .github/api_best_practices.md:17]: I feel the suggestion is leaking implementation details and we should refrain from doing so at higher level. Therefore, if some good engineer implements a `XYZService` interface then the argument fails.

Nonetheless, there exists a service of type `APIServiceWithCustomRoutes` registered for each custom route, so there exists a grpc endpoint.

File: .github/api_best_practices.md (hunk 3 above)
Comment (md2119) [reply .github/api_best_practices.md]: I feel it impacts everything below, so I will leave it as is for now.

File: .github/api_best_practices.md (hunk 6 above)
Comment (md2119) [reply .github/api_best_practices.md:30]: fair enough

File: .github/api_best_practices.md [hunk 8]
```diff
@@ -0,0 +1,231 @@
+# ACS API Guidelines
+An overview of the set of standardized guidelines and practices to ensure that the APIs across different 
+projects and teams adhere to a common structure and design. This consistency simplifies the process of understanding 
+and maintaining APIs. The guidelines helps identify potential pitfalls and design choices that could lead to issues to
```
Comment (kylape) [inline .github/api_best_practices.md:4]:
```suggestion
and maintaining APIs. These guidelines help identify potential pitfalls and design choices that could lead to issues to 
```

File: .github/api_best_practices.md [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+  - [gRPC Method Name](#grpc-method-name)
+  - [gRPC Message Name](#grpc-message-name)
+
+## Naming Guidelines
+
+### gRPC Service Name
+Service name **must** be unique and should use a noun that generally refers to a resource or product component 
+e.g. `DeploymentService`, `ReportService`, `ComplianceService`. Intuitive and well-known short forms or 
+abbreviations may be used in some cases (and could even be preferable) for succinctness 
+e.g. `ReportConfigService`, `RbacService`.
```
Comment (kylape) [inline .github/api_best_practices.md:21]: Not sure if this is too specific, but should we define case conventions? e.g. Should it be `RbacService` or `RBACService`?  (Sorry if this is some implied gRPC convention already.)

File: .github/api_best_practices.md [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+
+- [Naming Guidelines](#naming-guidelines)
+  - [gRPC Service Name](#grpc-service-name)
+  - [gRPC Method Name](#grpc-method-name)
+  - [gRPC Message Name](#grpc-message-name)
+
+## Naming Guidelines
+
+### gRPC Service Name
+Service name **must** be unique and should use a noun that generally refers to a resource or product component
```
Comment (kylape) [inline .github/api_best_practices.md:18]:
```suggestion
The service name **must** be unique and should use a noun that generally refers to a resource or product component 
```

File: .github/api_best_practices.md [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+
+### gRPC Service Name
+Service name **must** be unique and should use a noun that generally refers to a resource or product component 
+e.g. `DeploymentService`, `ReportService`, `ComplianceService`. Intuitive and well-known short forms or 
+abbreviations may be used in some cases (and could even be preferable) for succinctness 
+e.g. `ReportConfigService`, `RbacService`. 
+All gRPC methods grouped into a single service must generally pertain to the primary resource of the service.
+
+Note:
+- All service defined in versioned package [/proto/api/v2](https://github.com/stackrox/stackrox/blob/master/proto/v2)
```
Comment (kylape) [inline .github/api_best_practices.md:25]:
```suggestion
- All services defined in versioned package [/proto/api/v2](https://github.com/stackrox/stackrox/blob/master/proto/v2)
```

File: .github/api_best_practices.md [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+## Naming Guidelines
+
+### gRPC Service Name
+Service name **must** be unique and should use a noun that generally refers to a resource or product component 
+e.g. `DeploymentService`, `ReportService`, `ComplianceService`. Intuitive and well-known short forms or 
+abbreviations may be used in some cases (and could even be preferable) for succinctness 
+e.g. `ReportConfigService`, `RbacService`. 
+All gRPC methods grouped into a single service must generally pertain to the primary resource of the service.
+
+Note:
```
Comment (kylape) [inline .github/api_best_practices.md:24]: This doesn't seem to fit under the naming section.  Should it instead be its own section?  If there is going to be another PR about API versioning, perhaps this note would fit better in that PR.

File: .github/api_best_practices.md [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+**must not** import from [/proto/storage](https://github.com/stackrox/stackrox/blob/master/proto/storage) and
+[/proto/internalapi](https://github.com/stackrox/stackrox/blob/master/proto/internalapi) packages.
+- All new gRPC methods defined in versioned package [/proto/api/v1](https://github.com/stackrox/stackrox/blob/master/proto/v1) 
+**must not** import from `/proto/storage` and `/proto/internalapi` packages.
+- All services of the type `APIServiceWithCustomRoutes` **must not** import from `/proto/storage` and `/proto/internalapi` packages.
+
+### gRPC Method Name
+Methods should be named such that they provide insights into the functionality. Typically, the method 
+name should follow the _VerbNoun_ convention. For example, `StartComplianceScan`, `RunComplianceScan`, 
+and `GetComplianceScan` are not the same. `StartComplianceScan` must return without waiting for the compliance
```
Comment (kylape) [inline .github/api_best_practices.md:35]: Do we have a standard verb for a method that starts a task and _does_ wait for it to complete?

File: .github/api_best_practices.md [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+| Verb   | Noun           | Method name         |
+|--------|----------------|---------------------|
+| List   | Deployment     | `ListDeployments`   |
+| Get    | Deployment     | `GetDeployment`     |
+| Update | Deployment     | `UpdateDeployment`  |
+| Delete | Deployment     | `DeleteDeployment`  |
+| Notify | Violation      | `NotifyViolation`   |
+| Run    | ComplianceScan | `RunComplianceScan` |
+
+It is **recommended** that the verbs be imperative instead of inquisitive. Generally, the noun should be the resource type.
```
Comment (kylape) [inline .github/api_best_practices.md:49]: Do you intend to be consistent with the use of bolding to indicate the strength of the suggestion?  i.e. recommended, should, may, must.  I see various places where these words are not bolded.

File: .github/api_best_practices.md [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+| Verb                      | Noun           | Method name         | Request message            | Response message            |
+|---------------------------|----------------|---------------------|----------------------------|-----------------------------|
+| List                      | Deployment     | `ListDeployments`   | `ListDeploymentRequest`    | `ListDeploymentResponse`    |
+| Get                       | Deployment     | `GetDeployment`     | `GetDeploymentRequest`     | `GetDeploymentResponse`     |   
+| Update                    | Deployment     | `UpdateDeployment`  | `UpdateDeploymentRequest`  | `UpdateDeploymentResponse`  |  
+| Delete                    | Deployment     | `DeleteDeployment`  | `DeleteDeploymentRequest`  | `google.protobuf.Empty`     |     
+| Get                       | ReportStatus   | `GetReportStatus`   | `GetReportStatusRequest`   | `GetReportStatusResponse`   |  
+| Run                       | ComplianceScan | `RunComplianceScan` | `RunComplianceScanRequest` | `RunComplianceScanResponse` | 
+
+Avoid prepositions as much as possible (e.g. “for”, “with”; `DeploymentWithProcessInfo`, `DeploymentWithImageScan`).
```
Comment (kylape) [inline .github/api_best_practices.md:162]: Given the imperative above to have messages based on their corresponding service name, do we need to repeat the requirement to avoid prepositions?  It seems like this is accomplished if the requirement is followed for service naming conventions alone.

**Review by kylape — COMMENTED** [review]
Overall I think these guidelines are reasonable and easy to understand.

File: .github/api_best_practices.md (hunk 9 above)
Comment (md2119) [reply .github/api_best_practices.md:21]: It seems too specific to me. We have to let engineers use their best judgment.

File: .github/api_best_practices.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (md2119) [reply .github/api_best_practices.md:18]: Sticking to a standard naming convention when implementing services is a reasonable ask.

File: .github/api_best_practices.md (hunk 15 above)
Comment (md2119) [reply .github/api_best_practices.md:162]: It is good to repeat, it helps with retention :)

File: .github/api_best_practices.md (hunk 13 above)
Comment (md2119) [reply .github/api_best_practices.md:35]: I think that should be accomplished through the request option. For example, `RunComplianceScan` is ambiguous, however, if its request is:
```
RunComplianceScanRequest {
  bool wait_for_scan_completion
} 
```

then the communication is clear.

Comment (kylape) [reply .github/api_best_practices.md:35]: That looks reasonable to me.  Can that be integrated into this paragraph somehow? e.g. "the method name is ambiguous and should be avoided unless fields in the message help clarify the meaning."

File: .github/api_best_practices.md (hunk 11 above)
Comment (md2119) [reply .github/api_best_practices.md:25]: irrelevant because the note has been removed

**Review by msugakov — APPROVED** [review]
Good from my side. I recommend renaming the file per comment.

**Review by kylape — APPROVED** [review]

File: .github/api_best_practices.md [hunk 16]
```diff
@@ -0,0 +1,231 @@
+# ACS API Guidelines
```
[the remark this answers is not in the document]
Comment (md2119) [reply .github/api_best_practices.md:1]: I will address your comment in a follow-up PR

File: .github/api_best_practices.md [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+</td>
+<td>
+
+```
+GenerateDeploymentNetworkPolicy
+
+GenerateDeploymentNetworkPolicyRequest {
+  bool from_baseline;  
+  bool from_network_flows;
+}
```
Comment (rukletsov) [inline .github/api_best_practices.md:81]: While I get what you try to illustrate here (don't litter the URL, use query or body instead), all aspects of examples in guidelines are important. I doubt we want to encourage use of bool fields, so I recommend (also note correct formatting and proper proto syntax):
```
message GenerateDeploymentNetworkPolicyRequest {
    enum Source {
        UNKNOWN       = 0;
        BASELINE      = 1;
        NETWORK_FLOWS = 2;
    }

    Source from = 1;
}
```

File: .github/api_best_practices.md [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+All fields in the message **must** be lowercase and underscore separated names. The JSON names for the fields are 
+autogenerated by the proto compiler. By default, field names are converted to camel case notation.
+
+| Proto field name           | JSON field name        |
+|----------------------------|------------------------|
+| `network_data_start_time`  | `networkDataStartTime` |
+| `expiry_date`              | `expiryDate`           |
+
+Be explicit about conveying the specific purpose of fields e.g. instead of `expires_on`
+use `expiry_date`(/`timestamp`), instead of `network_data_since` use `network_data_start_time`.
```
Comment (rukletsov) [inline .github/api_best_practices.md:231]: What is the motivation for this one? I actually think that the suggested names are longer and not really more informative and hence inferior.

File: .github/api_best_practices.md [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+Methods (and URLs) should be named such that they provide insights into the functionality. Typically, the method 
+name should follow the _VerbNoun_ convention. For example, `StartComplianceScan`, `RunComplianceScan`, 
+and `GetComplianceScan` are not the same. `StartComplianceScan` must return without waiting for the compliance 
+scan to complete. `RunComplianceScan` may or may not wait for a compliance scan; the method name is ambiguous 
+and should be avoided if it starts a process and returns without waiting. `GetComplianceScan` should not run 
+a compliance scan but only fetches a stored one.
+
+| Verb   | Noun           | Method name         |
+|--------|----------------|---------------------|
+| List   | Deployment     | `ListDeployments`   |
```
[the remark this answers is not in the document]
Comment (md2119) [reply .github/api_best_practices.md:42]: Yes, it will be addressed in one of the upcoming PRs.

File: .github/api_best_practices.md (hunk 17 above)
Comment (md2119) [reply .github/api_best_practices.md:81]: I specifically did not choose the enum path because it has the limitation that it does not provide setting multiple options.  For example, a network policy generation can consider both baseline and current net pols. This is where graphQL has a clear advantage.

File: .github/api_best_practices.md (hunk 18 above)
Comment (md2119) [reply .github/api_best_practices.md:231]: Clarity is preferred. For example, `network_data_since` does not convey the unit; it may or may not include time portion.
